### PR TITLE
Better errors and memory management

### PR DESCRIPTION
**Pull request by asquared31415**
Changes some `.unwrap()` into `.expect(reason)` to clarify the source of the error
Adds additional error handling in some locations
Properly handles allocation errors with `handle_alloc_error`
Does not allocate in EntityBuilder when not needed

### CODE REVIEW

File: src/entities.rs [hunk 1]
```diff
@@ -63,12 +76,28 @@ impl Entities {
     }
 
     pub fn is_alive(&self, entity: EcsId) -> bool {
-        *self.generations.get(entity.uindex()).unwrap() == entity.generation()
+        let generation = entity.generation();
```
Comment (BoxyUwU) [inline src/entities.rs]: Can you move this to above the match block since its not used until then

File: src/entity_builder.rs [hunk 2]
```diff
@@ -72,14 +71,14 @@ impl<'a> EntityBuilder<'a> {
...
-        let data =
-            NonNull::new(unsafe { alloc(std::alloc::Layout::from_size_align(cap, 1).unwrap()) })
-                .expect("Failed to allocate for EntityBuilder");
+        let layout = std::alloc::Layout::from_size_align(cap, 1).unwrap();
+        let data = unsafe { alloc(layout) };
 
         Self {
-            data,
-            len: 0,
+            data: NonNull::new(data).unwrap_or_else(|| handle_alloc_error(layout)),
```
Comment (BoxyUwU) [inline src/entity_builder.rs]: Can you move this to a local

File: src/entity_builder.rs [hunk 3]
```diff
@@ -92,18 +91,32 @@ impl<'a> EntityBuilder<'a> {
...
+            let layout = std::alloc::Layout::from_size_align(self.cap, 1).unwrap();
+            let new_ptr = unsafe {
+                realloc(
+                    // self.data will never be null here
+                    self.data.as_ptr(),
+                    layout,
+                    new_size,
+                )
+            };
+            self.data = NonNull::new(new_ptr).unwrap_or_else(|| handle_alloc_error(layout));
```
Comment (BoxyUwU) [inline src/entity_builder.rs:117]: This is using a ``layout`` with the size set to the old capacity not a layout with the size of new_size

File: src/entity_builder.rs [hunk 4]
```diff
@@ -92,18 +91,32 @@ impl<'a> EntityBuilder<'a> {
...
 
-        let new_ptr = unsafe {
-            realloc(
-                self.data.as_ptr(),
-                std::alloc::Layout::from_size_align(self.cap, 1).unwrap(),
-                new_size,
-            )
-        };
+        if self.cap == 0 {
+            self.cap = new_size;
```
Comment (BoxyUwU) [inline src/entity_builder.rs]: We should move this self.cap assignment to after we set self.data to the new pointer

**Review by BoxyUwU — CHANGES_REQUESTED** [review]

**Review by BoxyUwU — APPROVED** [review]

**Comment by asquared31415** [comment]
I think this is ready now

**Comment by BoxyUwU** [comment]
:) thank u

**Pull request merged**